### PR TITLE
automation, exim: Fix generated AF templates

### PR DESCRIPTION
**Pull request by ricekot**
## Overview
Some AF templates do not end with an empty line (e.g. [`passiveScan-config-max.yaml`](https://github.com/zaproxy/zap-extensions/blob/f53114fa7b11453f2ed4c641ac86710e4c90d9cd/addOns/pscan/src/main/resources/org/zaproxy/addon/pscan/automation/jobs/passiveScan-config-max.yaml)) and result in invalid templates being generated:

image

This is fixed by normalizing the trailing newlines at the time of template generation in the automation add-on. Also, deprecated jobs and data jobs are skipped, and jobs are sorted by their types.

The indentation in exim templates was not correct and was also adjusted.

## Checklist
- [ ] Update help
- [x] Update changelog
- [x] Run `./gradlew spotlessApply` for code formatting
- [ ] Write tests
- [x] Check code coverage
- [x] Sign-off commits
- [x] Squash commits
- [x] Use a descriptive title

### CODE REVIEW

**Review by psiinon — APPROVED** [review]

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/ExtensionAutomation.java [hunk 1]
```diff
@@ -287,15 +290,21 @@ public void generateTemplateFile(String filename, boolean incAll) {
...
-                                    } else {
-                                        fw.write(j.getTemplateDataMin());
+                                    String template =
+                                            incAll
+                                                    ? j.getTemplateDataMax()
+                                                    : j.getTemplateDataMin();
+                                    if (StringUtils.isBlank(template)) {
+                                        return;
                                     }
+                                    template = template.stripTrailing() + "\n\n";
```
Comment (kingthorin) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/ExtensionAutomation.java:306]: Doesn't this need to account for windows users manually editing the templates?

Comment (thc202) [reply addOns/automation/src/main/java/org/zaproxy/addon/automation/ExtensionAutomation.java:306]: They are bundled in the add-ons.

**Review by kingthorin — APPROVED** [review]

File: addOns/exim/src/main/resources/org/zaproxy/addon/exim/resources/export-max.yaml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-- type: export            # Exports data into a file
```
Comment (thc202) [inline addOns/exim/src/main/resources/org/zaproxy/addon/exim/resources/export-max.yaml:1]: Worth updating the tests, e.g.:
```diff
--- a/addOns/exim/src/test/java/org/zaproxy/addon/exim/automation/ExportJobUnitTest.java
+++ b/addOns/exim/src/test/java/org/zaproxy/addon/exim/automation/ExportJobUnitTest.java
@@ -79,12 +79,18 @@ class ExportJobUnitTest extends TestUtils {
         assertThat(job.getType(), is(equalTo("export")));
         assertThat(job.getName(), is(equalTo("export")));
         assertThat(job.getOrder(), is(equalTo(AutomationJob.Order.AFTER_ATTACK)));
-        assertThat(job.getTemplateDataMin(), is(not(equalTo(""))));
-        assertThat(job.getTemplateDataMax(), is(not(equalTo(""))));
+        assertValidTemplate(job.getTemplateDataMin());
+        assertValidTemplate(job.getTemplateDataMax());
         assertThat(job.getParamMethodObject(), is(nullValue()));
         assertThat(job.getParamMethodName(), is(nullValue()));
     }
 
+    private static void assertValidTemplate(String value) {
+        assertThat(value, is(not(equalTo(""))));
+        new Yaml().load(value);
+    }
+
```

Comment (kingthorin) [reply addOns/exim/src/main/resources/org/zaproxy/addon/exim/resources/export-max.yaml:1]: Should the "new Yaml" bit assert does not throw?

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/ExtensionAutomation.java [hunk 3]
```diff
@@ -287,15 +290,21 @@ public void generateTemplateFile(String filename, boolean incAll) {
             }
 
             jobs.values().stream()
-                    .sorted()
+                    .filter(job -> job.getClass().getAnnotation(Deprecated.class) == null)
+                    .filter(Predicate.not(AutomationJob::isDataJob))
+                    .sorted(Comparator.comparing(AutomationJob::getType))
```
Comment (thc202) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/ExtensionAutomation.java:295]: Probably better to sort them first by `Order` as they would naturally be in the right position overall.

File: addOns/exim/src/main/resources/org/zaproxy/addon/exim/resources/export-max.yaml (hunk 2 above)
Comment (thc202) [reply addOns/exim/src/main/resources/org/zaproxy/addon/exim/resources/export-max.yaml:1]: Yeah, that's better (the outcome is the same).

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Pull request merged**